File: app/src/main/java/igrek/songbook/service/filesystem/ExternalCardService.java
package igrek.songbook.service.filesystem;


import android.app.Activity;
import android.os.Build;
import android.os.Environment;
import android.support.annotation.NonNull;

import com.google.common.base.Joiner;

import java.io.File;
import java.io.InputStream;
import java.util.HashSet;
import java.util.Locale;

import javax.inject.Inject;

import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.logger.Logger;
import igrek.songbook.logger.LoggerFactory;

/**
 * service to find external sd card location (it's not so obvious)
 */
public class ExternalCardService {
	
	private Logger logger = LoggerFactory.getLogger();
	@Inject
	Activity activity;
	
	private String externalSDPath;
	
	public ExternalCardService() {
		DaggerIoc.getFactoryComponent().inject(this);
		logger.debug(Joiner.on(", ").join(getExternalMounts()));
		externalSDPath = findExternalSDPath();
		logger.debug("External SD Card path: " + externalSDPath);
		//logger.debug("DEVICE = " + android.os.Build.DEVICE);
		//logger.debug("MANUFACTURER = " + android.os.Build.MANUFACTURER);
	}
	
	public String getExternalSDPath() {
		return externalSDPath;
	}
	
	protected String findExternalSDPath() {
		return new FirstRuleChecker<String>().addRule(this::isSamsung, () -> checkDirExists("/storage/extSdCard"))
				.addRule(() -> getExternalMount())
				.addRule(() -> checkDirExists("/storage/extSdCard"))
				.addRule(() -> checkDirExists("/storage/external_sd"))
				.addRule(() -> checkDirExists("/storage/ext_sd"))
				.addRule(() -> checkDirExists("/storage/external"))
				.addRule(() -> getExternalStorageDirectory())
				.find();
	}
	
	@NonNull
	private String getExternalStorageDirectory() {
		logger.warn("getting Environment.getExternalStorageDirectory() -  it's probably not what it's named for");
		return Environment.getExternalStorageDirectory().getAbsolutePath();
	}
	
	private String checkDirExists(String path) {
		File f = new File(path);
		if (f.exists() && f.isDirectory())
			return path;
		return null;
	}
	
	private boolean isSamsung() {
		return Build.DEVICE.contains("samsung") || Build.MANUFACTURER.contains("samsung");
	}
	
	private String getExternalMount() {
		HashSet<String> externalMounts = getExternalMounts();
		if (externalMounts.size() > 1) {
			logger.warn("multiple external mounts found, getting the first one");
		}
		if (!externalMounts.isEmpty())
			return externalMounts.iterator().next();
		return null;
	}
	
	private HashSet<String> getExternalMounts() {
		final HashSet<String> out = new HashSet<>();
		String reg = "(?i).*vold.*(vfat|ntfs|exfat|fat32|ext3|ext4).*rw.*";
		String s = "";
		try {
			final Process process = new ProcessBuilder().command("mount")
					.redirectErrorStream(true)
					.start();
			process.waitFor();
			final InputStream is = process.getInputStream();
			final byte[] buffer = new byte[1024];
			while (is.read(buffer) != -1) {
				s = s + new String(buffer);
			}
			is.close();
		} catch (final Exception e) {
			e.printStackTrace();
		}
		
		// parse output
		final String[] lines = s.split("\n");
		for (String line : lines) {
			if (!line.toLowerCase(Locale.US).contains("asec")) {
				if (line.matches(reg)) {
					String[] parts = line.split(" ");
					for (String part : parts) {
						if (part.startsWith("/"))
							if (!part.toLowerCase(Locale.US).contains("vold")) {
								// lg workaround - brilliant
								if (android.os.Build.MANUFACTURER.contains("LGE"))
									part = part.replaceAll("^/mnt/media_rw", "/storage");
								out.add(part);
							}
					}
				}
			}
		}
		return out;
	}
	
}

File: app/src/main/java/igrek/songbook/system/filesystem/ExternalCardService.kt
package igrek.songbook.system.filesystem


import android.app.Activity
import android.os.Build
import android.os.Environment
import igrek.songbook.info.logger.LoggerFactory
import java.io.File
import java.util.*

/**
 * service to find external sd card location (it's not so obvious)
 */
class ExternalCardService {

    private val logger = LoggerFactory.logger
    private val externalSDPath: String?

    private val externalStorageDirectory: String
        get() {
            logger.warn("getting Environment.getExternalStorageDirectory() -  it's probably not what it's named for")
            return Environment.getExternalStorageDirectory().absolutePath
        }

    private val isSamsung: Boolean
        get() = Build.DEVICE.contains("samsung") || Build.MANUFACTURER.contains("samsung")

    private val externalMount: String?
        get() {
            val externalMounts = externalMounts
            if (externalMounts.size > 1) {
                logger.warn("multiple external mounts found, getting the first one")
            }
            return if (externalMounts.isNotEmpty()) externalMounts.iterator().next() else null
        }

    private// parse output
    // lg workaround - brilliant
    val externalMounts: HashSet<String>
        get() {
            val out = HashSet<String>()
            val reg = "(?i).*vold.*(vfat|ntfs|exfat|fat32|ext3|ext4).*rw.*"
            val s = StringBuilder()
            try {
                val process = ProcessBuilder().command("mount")
                        .redirectErrorStream(true)
                        .start()
                process.waitFor()
                val `is` = process.inputStream
                val buffer = ByteArray(1024)
                while (`is`.read(buffer) != -1) {
                    s.append(String(buffer))
                }
                `is`.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }

            val lines = s.toString().split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (line in lines) {
                if (!line.toLowerCase(Locale.US).contains("asec")) {
                    if (line.matches(reg.toRegex())) {
                        val parts = line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                        for (part in parts) {
                            if (part.startsWith("/"))
                                if (!part.toLowerCase(Locale.US).contains("vold")) {
                                    var partr = part
                                    if (Build.MANUFACTURER.contains("LGE"))
                                        partr = part.replace("^/mnt/media_rw".toRegex(), "/storage")
                                    out.add(partr)
                                }
                        }
                    }
                }
            }
            return out
        }

    init {
        externalSDPath = findExternalSDPath()
        logger.debug("External SD Card path detected: " + externalSDPath!!)
    }

    private fun findExternalSDPath(): String? {
        return FirstRuleChecker<String>()
                .addRule({ this.isSamsung }, { checkDirExists("/storage/extSdCard") })
                .addRule { this.externalMount }
                .addRule { checkDirExists("/storage/extSdCard") }
                .addRule { checkDirExists("/storage/external_sd") }
                .addRule { checkDirExists("/storage/ext_sd") }
                .addRule { checkDirExists("/storage/external") }
                .addRule { this.externalStorageDirectory }
                .find()
    }

    private fun checkDirExists(path: String): String? {
        val f = File(path)
        return if (f.exists() && f.isDirectory) path else null
    }

}
